Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -384,4 +356,269 @@ static void _GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameSi
+                return fontNames;
+            }
+
+            [fontNames addObject:name];
+        }
+    }
+
+    return fontNames;
+}
+
+NSString* _DWriteGetFamilyNameForFontName(CFStringRef fontName) {
+    static NSDictionary<NSString*, NSString*>* fontToFamilyMap = nil;
+
+    static dispatch_once_t initOnce = 0;
+    dispatch_once(&initOnce,
+                  ^{
+                      // initialize fontToFamilyMap
+                      NSMutableDictionary<NSString*, NSString*>* initMap = [NSMutableDictionary<NSString*, NSString*> dictionary];
+                      NSArray<NSString*>* familyNames = _DWriteGetFontFamilyNames();
+
+                      for (NSString* familyName in familyNames) {
+                          NSArray<NSString*>* fontNames = _DWriteGetFontNamesForFamilyName(familyName);
+
+                          for (NSString* fontName in fontNames) {
+                              [initMap setObject:familyName forKey:fontName];
+                          }
+                      }
+
+                      fontToFamilyMap = [initMap copy];
+                  });
+
+    return [fontToFamilyMap objectForKey:(__bridge NSString*)fontName];
+}
+
+// Private helper that acquires an IDWriteFontFamily object for a given family name
+static ComPtr<IDWriteFontFamily> __GetDWriteFontFamily(NSString* familyName) {
+    ComPtr<IDWriteFontFamily> fontFamily;
+
+    ComPtr<IDWriteFactory> factory = __GetDWriteFactoryInstance();
+
+    ComPtr<IDWriteFontCollection> systemFontCollection;
+    THROW_IF_FAILED(factory->GetSystemFontCollection(&systemFontCollection));
+
+    NSUInteger familyNameLength = familyName.length;
+    std::vector<UniChar> unicharFamilyName(familyNameLength + 1);
+    [familyName getCharacters:unicharFamilyName.data() range:NSMakeRange(0, familyNameLength)];
+
+    UINT32 fontFamilyIndex;
+    BOOL fontFamilyExists;
+
+    THROW_IF_FAILED(systemFontCollection->FindFamilyName(reinterpret_cast<const wchar_t*>(unicharFamilyName.data()),
+                                                         &fontFamilyIndex,
+                                                         &fontFamilyExists));
+
+    if (!fontFamilyExists) {
+        TraceError(TAG, L"Unable to find font family \"%ws\"", unicharFamilyName.data());
+        return fontFamily;
+    }
+
+    THROW_IF_FAILED(systemFontCollection->GetFontFamily(fontFamilyIndex, &fontFamily));
+    return fontFamily;
+}
+
+// Private helper that parses a font name, and returns appropriate weight, stretch, and style values for DWrite accordingly
+static void __InitDWriteFontPropertiesFromName(CFStringRef fontName,
+                                               DWRITE_FONT_WEIGHT* weight,
+                                               DWRITE_FONT_STRETCH* stretch,
+                                               DWRITE_FONT_STYLE* style) {
+    // Set some defaults for when weight/stretch/style are not mentioned in the name
+    *weight = DWRITE_FONT_WEIGHT_NORMAL;
```

Comment:
It's a file-private helper, we can assume that the pointers are good. I'll add a comment.

---
In reply to: [78207940](https://github.com/Microsoft/WinObjC/pull/987#discussion_r78207940) [](ancestors = 78207940)